src/ember-app.js

```diff
@@ -245,16 +263,16 @@ class EmberApp {
    * per request turned on and a new request comes in.
    *
    */
-  async _getNewApplicationInstance() {
+  async getNewApplicationInstance() {
     let app;
 
-    if (this._pendingNewApplicationInstance) {
-      let pendingAppInstancePromise = this._pendingNewApplicationInstance;
-      this._pendingNewApplicationInstance = undefined;
-      app = await pendingAppInstancePromise;
-    } else {
-      // if there is no current pending application instance, create a new one on-demand.
+    if (this._sandboxApplicationInstanceQueue.isEmpty()) {
```

Comment:
I'd prefer to always get the instance from the queue, and the queue would return a structured object that indicates if it had to "build on demand" or had a pre-warmed instance (related to comment about RE: passing the "builder" function into the `Queue` constructor).

Reply:
I am not sure I can do that unless I leak some details into the queue. The instance differs when it is returned from the queue and when it is built on demand. The instance in a queue is a promise which resolves to the sandbox while build on demand is an instance directly.

basically when it is pre-warmed we call `Promise.resolve()...` while on demand it is `buildApp()` which are two different functions.